cmd/localstack/awsutil.go

```diff
@@ -196,6 +201,178 @@ func DownloadCodeArchive(url string) {
+						if err != nil {
+							log.Errorln("Error watching folder: ", folder, err)
+						}
+					}
+				}
+				// remove in case of remove / rename (rename within the folder will trigger a separate create event)
+			} else if event.Has(fsnotify.Remove) || event.Has(fsnotify.Rename) {
+				// remove all file watchers if it is in our folders list
+				toBeRemovedDirs, newWatchedFolders := getSubFoldersInList(event.Name, c.watchedFolders)
+				c.watchedFolders = newWatchedFolders
+				for _, dir := range toBeRemovedDirs {
+					err := c.watcher.Remove(dir)
+					if err != nil {
+						log.Warnln("Error removing path: ", event.Name, err)
+					}
+				}
+			}
+			c.changeChannel <- event.Name
+		case err, ok := <-c.watcher.Errors():
+			if !ok {
+				log.Println("error:", err)
+				return
+			}
+			log.Println("error:", err)
+		}
+	}
+}
+
+func (c *ChangeListener) addTargetPaths(targetPaths []string) {
+	// Add all target paths and subfolders
+	for _, targetPath := range targetPaths {
+		subfolders := getSubFolders(targetPath)
+		log.Infoln("Subfolders: ", subfolders)
+		for _, target := range subfolders {
+			err := c.watcher.Add(target)
+			if err != nil {
+				log.Fatal(err)
+			}
+		}
+	}
+}
+
+func (c *ChangeListener) close() error {
+	return c.watcher.Close()
+}
+
+func resetListener(changeChannel <-chan bool, server *CustomInteropServer) {
+	for {
+		_, more := <-changeChannel
+		if !more {
+			return
+		}
+		log.Println("Resetting environment...")
+		_, err := server.Reset("HotReload", 2000)
+		if err != nil {
+			log.Warnln("Error resetting server: ", err)
+		}
+	}
+
+}
+
+func debounceChannel(changeChannel <-chan string, duration time.Duration) <-chan bool {
+	resultChannel := make(chan bool, 10)
+	// debouncer to limit restarts
+	timer := time.NewTimer(duration)
+	// immediately stop the timer, since we do not want to reload right at the startup
+	if !timer.Stop() {
+		// we have to drain the channel in case the timer already fired
+		<-timer.C
+	}
+	go func() {
+		for {
+			select {
+			case _, more := <-changeChannel:
+				if !more {
+					timer.Stop()
+					close(resultChannel)
+					return
+				}
+				timer.Reset(duration)
+			case <-timer.C:
+				resultChannel <- true
+			}
+		}
+	}()
+	return resultChannel
+}
+
+func RunHotReloadingListener(server *CustomInteropServer, targetPaths []string, opts *LsOpts, ctx context.Context) {
+	if !opts.HotReloading {
+		log.Debugln("Hot reloading disabled.")
+		return
+	}
+	defaultDuration := 500 * time.Millisecond
+	log.Infoln("Hot reloading enabled, starting filewatcher.", targetPaths)
+	changeListener, err := createChangeListener()
+	if err != nil {
+		log.Errorln("Hot reloading disabled due to change listener error.", err)
+		return
+	}
+	defer changeListener.close()
+	go changeListener.watch()
+	changeListener.addTargetPaths(targetPaths)
+	debouncedChannel := debounceChannel(changeListener.changeChannel, defaultDuration)
+	go resetListener(debouncedChannel, server)
```

Comment:
suggestion: Could also do the debouncing in the ChangeListener and add the debounce duration/interval to the struct. wdyt?

Reply:
That should help avoiding this complex call pattern here :thinking: